# Studying RBM20 mutations in mice

## Experiment descriptions

2 experiments were performed by Markus Grosch. The sequencing was performed using 
Illumina (bulk RNA-seq) from mouse hearts.

Two main mutations are studied in the RBM20 gene: P635L located at chr19:53831671 
(C wt \> T mut) and R636Q located at chr19:53831674/675 (GT wt \> AG mut).

### Induce mutation - experiment 1

Mutations were generated by CRISPR Cas9. Samples information are listed in 
config/samples_metadata.txt. There are 5 samples for each group: 
wild type (WT); heterozygote (HET); homozygote (HOM).

### After base editing - experiment 2

Multiple base editors were used. Samples information are listed in 
config/samples_afterbaseediting_metadata.txt.

## Analysis

The analysis was done using *Snakemake* [MÃ¶lder et al. 2021] 
(except for the *rMATS* analysis [Shen et al. 2014]) and followed the *Snakemake* 
recommended directory structure.

### Alignment

#### STAR

The alignment of the different samples is performed using *STAR* [Dobin et al. 2013]. 
We use the GENCODE mouse annotation version vM29 with the primary assembly GRCm39 genome. 
We create the indexes and then align the reads for each sample using the default options.

Script located at:

- workflow/rules/STAR_alignment.smk

### IVG visualization

We extract the sequences from the bam file to visualize the mutation's loci in 
*IGV* [Robinson et al. 2017].

Script located at:

- workflow/rules/extract_mutation.smk

### DESeq2

We use *DESeq2* [Love et al. 2019] to perform the differential expression analysis 
for the samples of the first experiment.

The first step is to create count matrices from the *.bam* files using the R 
package *Rsubread* [Liao et al. 2019]. The results are saved in rds object that 
can be load in the R script to perform the analysis.

We perform each comparison for each mutation/conditions:

- P635L: HET vs WT; HOM vs WT;
- R636Q: HET vs WT; HOM vs WT;
- P635L after base editing: Nterm_SpRY_and_Cterm_SpRY_gRNA5 vs Nterm_NRTH_Abe8e_and_Cterm_gRNA5;
Nterm_SpRY_and_Cterm_SpRY_gRNA5 vs PBS; Nterm_SpRY_and_Cterm_SpRY_gRNA5 vs none; 
Nterm_NRTH_Abe8e_and_Cterm_gRNA5 vs PBS; Nterm_NRTH_Abe8e_and_Cterm_gRNA5 vs none; 
PBS vs none;
- R636Q after base editing: Nterm_NRTH_Abe8e_and_Cterm_NRCH_gRNA6 vs PBS; 
Nterm_NRTH_Abe8e_and_Cterm_NRCH_gRNA6 vs none; PBS vs none.

We computed the log2 fold change (log2FC) per individual for each mutation, 
using the average per gene from the WT samples: *log2FC for gene A = log2(value of gene A / WT average for gene A)*.

For visualization, we plotted different heatmaps with the top genes expression 
as well as with the log2FC values.

Scripts located at:

- workflow/rules/DESeq_analysis.smk: to create the count matrices;
- workflow/scripts/DESeq_analysis_withSTARalignment.R: to perform the DESeq2 
analysis and plot all heatmaps.

### rMATS

We used *rMATS* [Shen et al. 2014] to detect differential alternative splicing events. 
Unfortunately, *rMATS* was not easily compatible at the time with snakemake. We then 
created its own conda environment that can be activated using *conda activate rMATS_env*.

All *rMATS* results are located outside the *results* directory, in its own directory:
*rmats_directory/*.

Scripts located at:

- rmats_directory/README.txt: list of the commands to run the *rMATS* comparisons. The input files used to perform the comparison were created manually (*mutation_condition/genotype_files.txt*);
- workflow/scripts/rmat_data_analysis.R: R scripts analyzing the output of *rMATS* python script.

## Appendix

### Kallisto Alignment

The initial alignment was performed using the *kallisto* mapper [Bray et al. 2016]. 
However, we then switched to STAR as it was a requirement for DEXseq.

Script located at:

- rules/kallisto_alignment.smk

### DEXseq

The initial splice junction events analysis was performed using *DEXseq* R package 
[Anders et al. 2012].

Script located at:

- rules/DEXSeq_analysis.smk
